sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/strategy/CanaryStrategy.java

```diff
@@ -1,60 +1,165 @@
+
+    public Collection<Step> canarySteps;
+
+    /**
+     * Creates a new Canary Strategy which will require 2 {@link #proceed()} calls from a user before following the
+     * provided {@code postCanaryStrategy}.
+     *
+     * @param postCanaryStrategy the strategy to use after the canary stage has completed
+     */
+    public CanaryStrategy(Strategy<Step> postCanaryStrategy) {
+        this(postCanaryStrategy, DEFAULT_PROCEED_COUNT);
+    }
+
+    /**
+     * Creates a new Canary Strategy which will require a specified number of {@link #proceed()} calls from a user
+     * before following the provided {@code postCanaryStrategy}.
+     *
+     * @param postCanaryStrategy the strategy to use after the canary stage has completed
+     * @param requiredProceeds the number of {@link #proceed()} calls to require before the provided strategy is
+     *     executed
+     */
+    public CanaryStrategy(Strategy<Step> postCanaryStrategy, int requiredProceeds) {
+        this.requiredProceeds = requiredProceeds;
+        this.strategy = postCanaryStrategy;
+        this.canarySteps = null;
+    }
 
     @Override
-    public Collection<C> getCandidates(Element<C> parentElement, Collection<String> dirtyAssets) {
-        if (!initialized.get()) {
-            children = parentElement.getChildren().stream()
-                    .filter(element -> (element.isPending() || element.isWaiting()))
+    public Collection<Step> getCandidates(Collection<Step> steps, Collection<String> dirtyAssets) {
+        if (canarySteps == null) {
+            // Create the N initial canary steps by marking them as interrupted.
+            canarySteps = steps.stream()
+                    .filter(step -> (step.isPending() || step.isInterrupted()))
+                    .limit(requiredProceeds)
                     .collect(Collectors.toList());
+            canarySteps.forEach(step -> step.interrupt());
+        }
 
-            for (int i = 0; i < 2 && i < children.size(); i++) {
-                children.get(i).getStrategy().interrupt();
-            }
-
-            initialized.set(true);
+        if (getNextCanaryStep() != null) {
+            // Still in canary. Only return subset of canary steps which are now eligible due to proceed() calls.
+            return canarySteps.stream()
+                    .filter(step -> PlanUtils.isEligibleCandidate(step, dirtyAssets))
+                    .collect(Collectors.toList());
         }
 
-        return super.getCandidates(parentElement, dirtyAssets);
+        // Canary period has ended. Revert to underlying strategy for the rest of the operation.
+        return strategy.getCandidates(steps, dirtyAssets);
+    }
+
+    @Override
+    public void interrupt() {
+        Step canaryStep = getNextCanaryStep();
+        if (canaryStep != null) {
+            // Ignoring interrupt as we are still in canary and therefore already interrupted.
+            return;
+        }
+        strategy.interrupt();
```

Comment:
If I interrupt the CanaryStrategy, before it calls getCandidates, it will interrupt its strategy.  Also, if first (at most two) steps are already proceeded, it will interrupt its strategy.